src/main/java/ch/njol/skript/lang/VariableString.java

```diff
@@ -527,48 +517,102 @@ private static ChatColor getLastColor(CharSequence s) {
 	public String toString() {
 		return toString(null, false);
 	}
-	
+
 	/**
-	 * Use {@link #toString(Event)} to get the actual string
+	 * Parses all expressions in the string and returns it.
+	 * If this is a simple string, the event may be null.
+	 * 
+	 * @param e Event to pass to the expressions.
+	 * @return The input string with all expressions replaced.
+	 */
+	public String toString(@Nullable Event event) {
+		if (isSimple) {
+			assert simple != null;
+			return simple;
+		}
+		if (event == null)
+			throw new IllegalArgumentException("Event may not be null in non-simple VariableStrings!");
+
+		Object[] string = this.string;
+		assert string != null;
+		StringBuilder builder = new StringBuilder();
+		List<Class<?>> types = new ArrayList<>();
+		for (Object object : string) {
+			if (object instanceof Expression<?>) {
+				Object[] objects = ((Expression<?>) object).getArray(event);
+				if (objects != null && objects.length > 0)
+					types.add(objects[0].getClass());
+				builder.append(Classes.toString(objects, true, mode));
+			} else {
+				builder.append(object);
+			}
+		}
+		String complete = builder.toString();
+		if (!types.isEmpty()) {
+			DefaultVariables data = script.getData(DefaultVariables.class);
+			if (data != null)
+				data.add(complete, types.toArray(new Class<?>[0]));
```

Comment:
I'm still unsure about the behavior here. My understanding is that the "hints"/add method here are for specifically tracking variable names/types, but VariableString is not restricted to solely variable names. Let me know if this *is* intended for non-variables, or if I am misunderstanding the purpose of this call.

Reply:
I don't know of an area of calling code where this would be impacted. Explain where you think this would be a problem.

Reply:
Added a check against the StringMode

Reply:
Yeah it was just my understanding that this check is intended for variables but a VariableString is not guaranteed to represent a variable :)